src/dispatch/experimental/durable/function.py

```diff
@@ -137,13 +137,13 @@ def __getstate__(self):
             if frame_state < FRAME_CLEARED:
                 print(f"IP = {ip}")
                 print(f"SP = {sp}")
-                for i, (is_null, value) in enumerate(stack):
+                for i, (is_null, value) in enumerate(stack if stack is not None else []):
                     if is_null:
                         print(f"stack[{i}] = NULL")
                     else:
                         print(f"stack[{i}] = {value}")
                 print(f"BP = {bp}")
-                for i, block in enumerate(blocks):
+                for i, block in enumerate(blocks if blocks is not None else []):
```

Comment:
These would have broken in the case where `stack` and `blocks` were `None`.

Reply:
Looks like mypy isn't smart enough to see that these are only `None` if `frame_state >= FRAME_CLEARED`, and that we're in a block where `frame_state < FRAME_CLEARED`.

Reply:
This analysis would be pretty difficult to make right? There are no constants in Python, `FRAME_CLEARED` could be altered concurrently by a thread, potentially in a module that imported this one.

Reply:
I don't think there are any points between checks where the GIL would be released and another thread allowed to run? There are calls to the C extension in there, but you have to explicitly [release](https://docs.python.org/3/c-api/init.html#releasing-the-gil-from-extension-code) the GIL from C. mypy either assumes that all C extension calls may release the GIL, or it just doesn't propagate constraints across blocks.

Reply:
Assumptions about when the GIL can be released seem like asking for trouble, the rules are a bit unclear as to when this may happen; I would expect mypy not to try to be that smart because it's a lot of efforts for unclear rewards (false positives make the program safer anyways).

We could verify by pulling the value into a local instead of depending on the global `FRAME_CLEARED` variable.

Reply:
Oh but even with a local, it doesn't know that `frame_state = ext.get_frame_state(g)` will return a value less than `FRAME_CLEARED` right?

Reply:
(I just tried with a local and it still complains that the values might be `None`)

Reply:
`stack` and `blocks` are set to `None` after `frame_state` is set, and if and only if `frame_state < FRAME_CLEARED`. The `frame_state` variable does not change after they're set to `None`.

I suspect the issue is that mypy forwards constraints _into_ blocks, but doesn't carry them _across_ blocks, because of potential path explosions. It seems to be smart enough to tighten types as you extend into `if` blocks, but in this case it doesn't carry the `frame_state < FRAME_CLEARED` constraint forward to an adjacent block.

No big deal, false positives are expected occasionally 🙂